### PR TITLE
Run ITs on devnet and testnet by default

### PR DESCRIPTION
**Pull request by mukulljangid**
Addresses #194 by running ITs against local, devnet and testnet as part of the build.

### CODE REVIEW

**Comment by mukulljangid** [comment]
Converting it to draft because `TicketBatch` amendment was enabled on altnet 03/30/22 and might take upto 2 weeks.

File: xrpl4j-integration-tests/src/test/java/org/xrpl/xrpl4j/tests/environment/XrplEnvironment.java [hunk 1]
```diff
@@ -36,16 +36,23 @@
...
-    logger.info("useTestnet set to {}, using {} for testing.",
-      isTestnetEnabled, isTestnetEnabled ? "testnet" : "local rippled"
-    );
-    return isTestnetEnabled ? new TestnetEnvironment() : new LocalRippledEnvironment();
+    // Use the testnet and devnet environment by default to run integration testing.
+    // Use -DuseLocal test to run integration tests on local rippled.
+    boolean isTestnetEnabled = System.getProperty("useTestnet").equals("true");
+    boolean isDevnetEnabled = System.getProperty("useDevnet").equals("true");
+    if (isTestnetEnabled) {
+      logger.info("Using testnet for integration testing.");
```
Comment (sappenin) [inline xrpl4j-integration-tests/src/test/java/org/xrpl/xrpl4j/tests/environment/XrplEnvironment.java:47]:
```suggestion
      logger.info("System property 'useTestnet' detected; Using Testnet for integration testing.");
```

File: xrpl4j-integration-tests/src/test/java/org/xrpl/xrpl4j/tests/environment/XrplEnvironment.java [hunk 2]
```diff
@@ -36,16 +36,23 @@
...
+      return new TestnetEnvironment();
+    } else if (isDevnetEnabled) {
+      logger.info("Using devnet for integration testing.");
```
Comment (sappenin) [inline xrpl4j-integration-tests/src/test/java/org/xrpl/xrpl4j/tests/environment/XrplEnvironment.java:50]:
```suggestion
    logger.info("System property 'useDevnet' detected; Using Devnet for integration testing.");
```

File: xrpl4j-integration-tests/src/test/java/org/xrpl/xrpl4j/tests/environment/XrplEnvironment.java [hunk 3]
```diff
@@ -36,16 +36,23 @@
...
+      return new DevnetEnvironment();
+    } else {
+      logger.info("Using local rippled for integration testing.");
```
Comment (sappenin) [inline xrpl4j-integration-tests/src/test/java/org/xrpl/xrpl4j/tests/environment/XrplEnvironment.java:53]:
```suggestion
      logger.info("Neither 'useTestNet' nor 'useDevnet' System properties detected. Using local rippled for integration testing.");
```

**Review by sappenin — CHANGES_REQUESTED** [review]

**Review by sappenin — APPROVED** [review]

File: pom.xml [hunk 4]
```diff
@@ -386,11 +387,47 @@
           </configuration>
           <executions>
             <execution>
-              <id>integration-tests</id>
+              <id>integration-tests-altnet</id>
```
Comment (sappenin) [inline pom.xml:390]: Let's change this to align with the naming used in code (also, most humans will call this `testnet` IMHO).

```suggestion
              <id>integration-tests-testnet</id>
```

**Review by sappenin — CHANGES_REQUESTED** [review]

**Review by sappenin — APPROVED** [review]

File: pom.xml [hunk 5]
```diff
@@ -377,10 +376,6 @@
           <groupId>org.apache.maven.plugins</groupId>
           <artifactId>maven-failsafe-plugin</artifactId>
           <version>2.22.2</version>
-          <configuration>
-            <skipITs>${skipITs}</skipITs>
-            <argLine>${failsafe.argLine}</argLine>
-          </configuration>
```
Comment (nkramer44) [inline pom.xml:383]: Why did we get rid of this? Is it not needed?

File: pom.xml [hunk 6]
```diff
@@ -301,8 +301,7 @@
     <checkstyle.config.location>checkstyle.xml</checkstyle.config.location>
     <checkstyle.violationSeverity>warning</checkstyle.violationSeverity>
 
-    <skipITs>false</skipITs>
-
+    <useLocal>false</useLocal>
```
Comment (nkramer44) [inline pom.xml:304]: Where does this get used?

File: xrpl4j-integration-tests/src/test/java/org/xrpl/xrpl4j/tests/AbstractIT.java [hunk 7]
```diff
@@ -368,4 +376,16 @@ public void sendIssuedCurrency(
...
 
   }
+
+  /**
+   * Helper function to print log statements for Integration Tests which is network specific.
+   *
+   * @param txName {@link TransactionType} to be logged for the executed transaction.
+   * @param hash   {@link Hash256} to be logged for the executed transaction.
+   */
+  public void logInfo(TransactionType txName, Hash256 hash) {
```
Comment (nkramer44) [inline xrpl4j-integration-tests/src/test/java/org/xrpl/xrpl4j/tests/AbstractIT.java:386]:
```suggestion
  public void logInfo(TransactionType transactionType, Hash256 hash) {
```

File: xrpl4j-integration-tests/src/test/java/org/xrpl/xrpl4j/tests/AbstractIT.java [hunk 8]
```diff
@@ -87,7 +88,12 @@ protected Wallet createRandomAccount() {
     // Create the account
     SeedWalletGenerationResult seedResult = walletFactory.randomWallet(true);
     final Wallet wallet = seedResult.wallet();
-    logger.info("Generated testnet wallet with XAddress={} (Classic={})", wallet.xAddress(), wallet.classicAddress());
+    String network = System.getProperty("useTestnet") != null ? "testnet" :
+      (System.getProperty("useDevnet") != null ? "devnet" : "local rippled");
+    logger.info(
+      "Generated {} wallet with XAddress={} (Classic={})",
+      network, wallet.xAddress(), wallet.classicAddress()
+    );
```
Comment (nkramer44) [inline xrpl4j-integration-tests/src/test/java/org/xrpl/xrpl4j/tests/AbstractIT.java:96]: What if we added a couple of methods in `XrplEnvironment`, one that returns the name (like `"local rippled"` or `"testnet"` or `"devnet"`), and one that returns the explorer URL to use. Each implementation of `XrplEnvironment` will easily know that information, and then you don't have to do all this switching on `System.getProperty`

File: xrpl4j-integration-tests/src/test/java/org/xrpl/xrpl4j/tests/EscrowIT.java [hunk 9]
```diff
@@ -66,7 +67,7 @@ public void createAndFinishTimeBasedEscrow() throws JsonRpcClientErrorException,
     EscrowCreate escrowCreate = EscrowCreate.builder()
       .account(senderWallet.classicAddress())
       .sequence(senderAccountInfo.accountData().sequence())
-      .fee(feeResult.drops().openLedgerFee())
+      .fee(XrpCurrencyAmount.ofXrp(new BigDecimal(1)))
```
Comment (nkramer44) [inline xrpl4j-integration-tests/src/test/java/org/xrpl/xrpl4j/tests/EscrowIT.java:70]: Is this just 1XRP so our transactions always go through?

**Review by nkramer44 — CHANGES_REQUESTED** [review]

File: pom.xml (hunk 5 above)
Comment (sappenin) [reply pom.xml:383]: The ITs were running even if '-DskipTests' was specified, which was a negatively unexpected outcome. Normally, the failsafe plugin will honor that first and then honor '-DskipITs' second, but the way the plugin was configured broke that connection so my ITs were _always_ running despite using common build idioms.

Also, the only thing dictating which environment to hit is the environment variable as opposed to anything related to the failsafe plugin, and the massive failsafe configured didn't seem to be adding any value from my perspective, so I thought that reducing that and just using env values in our builds was a simpler and cleaner twiddle than having a bunch of failsafe configs.

Any objections?

It's occurring to me that we might need to update the GH actions here though.

File: pom.xml (hunk 6 above)
Comment (sappenin) [reply pom.xml:304]: Doesn't seem to be used.

File: xrpl4j-integration-tests/src/test/java/org/xrpl/xrpl4j/tests/AbstractIT.java (hunk 8 above)
Comment (sappenin) [reply xrpl4j-integration-tests/src/test/java/org/xrpl/xrpl4j/tests/AbstractIT.java:96]: Agree - though I was thinking that would be implemented via https://github.com/XRPLF/xrpl4j/issues/266

File: xrpl4j-integration-tests/src/test/java/org/xrpl/xrpl4j/tests/EscrowIT.java (hunk 9 above)
Comment (sappenin) [reply xrpl4j-integration-tests/src/test/java/org/xrpl/xrpl4j/tests/EscrowIT.java:70]: @mukulljangid wdyt? Shouldn't we use feeutils here?

File: pom.xml (hunk 5 above)
Comment (nkramer44) [reply pom.xml:383]: Nope no objections. Just wanted to make this this wasn't an inadvertent deletion.

I think as long as ITs are running in CI, we're probably fine. I don't think we are skipping ITs in CI so this shouldn't affect that

File: xrpl4j-integration-tests/src/test/java/org/xrpl/xrpl4j/tests/EscrowIT.java (hunk 9 above)
Comment (mukulljangid) [reply xrpl4j-integration-tests/src/test/java/org/xrpl/xrpl4j/tests/EscrowIT.java:70]: Yess! Replacing it

File: xrpl4j-integration-tests/src/test/java/org/xrpl/xrpl4j/tests/AbstractIT.java (hunk 8 above)
Comment (mukulljangid) [reply xrpl4j-integration-tests/src/test/java/org/xrpl/xrpl4j/tests/AbstractIT.java:96]: Part of this has been accomplished by https://github.com/XRPLF/xrpl4j/pull/178/commits/d917186a4017d75c83260468d21cc3506590444c where it adds a Custom Environment. Should be pretty quick to create these envs once NFT is merged

File: pom.xml (hunk 6 above)
Comment (sappenin) [reply pom.xml:304]: Removed.

File: xrpl4j-integration-tests/src/test/java/org/xrpl/xrpl4j/tests/AbstractIT.java (hunk 8 above)
Comment (sappenin) [reply xrpl4j-integration-tests/src/test/java/org/xrpl/xrpl4j/tests/AbstractIT.java:96]: Resolving this for now. We'll address it when we merge in NFTs.

**Review by sappenin — APPROVED** [review]

**Pull request merged**